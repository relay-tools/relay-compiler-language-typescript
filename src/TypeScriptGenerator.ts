import { TypeGenerator, IRTransforms } from "relay-compiler";

import * as ts from "typescript";
import * as path from "path";

import {
  transformScalarType,
  transformInputType,
  ScalarTypeMapping,
  State
} from "./TypeScriptTypeTransformers";
import {
  GraphQLNonNull,
  GraphQLEnumType,
  GraphQLType,
  GraphQLScalarType,
  GraphQLNamedType
} from "graphql";
import {
  IRVisitor,
  SchemaUtils,
  IRTransform,
  Fragment,
  Root
} from "relay-compiler/lib/GraphQLCompilerPublic";

import { TypeGeneratorOptions } from "relay-compiler";

const { isAbstractType } = SchemaUtils;

export const generate: TypeGenerator["generate"] = (node, options) => {
  const ast: ts.Statement[] = IRVisitor.visit(node, createVisitor(options));
  const printer = ts.createPrinter({
    newLine: ts.NewLineKind.LineFeed
  });
  const resultFile = ts.createSourceFile(
    "grapghql-def.ts",
    "",
    ts.ScriptTarget.Latest,
    /*setParentNodes*/ false,
    ts.ScriptKind.TS
  );
  const fullProgramAst = ts.updateSourceFileNode(resultFile, ast);
  return printer.printNode(ts.EmitHint.SourceFile, fullProgramAst, resultFile);
};

type Selection = {
  key: string;
  schemaName?: string;
  value?: any;
  nodeType?: any;
  conditional?: boolean;
  concreteType?: string;
  ref?: string;
  nodeSelections?: SelectionMap | null;
};
type SelectionMap = Map<string, Selection>;

function nullthrows<T>(obj: T | null | undefined): T {
  if (obj == null) {
    throw new Error("Obj is null");
  }
  return obj;
}

function makeProp(
  selection: Selection,
  state: State,
  concreteType?: string
): ts.PropertySignature {
  let {
    key,
    schemaName,
    value,
    conditional,
    nodeType,
    nodeSelections
  } = selection;
  if (nodeType) {
    value = transformScalarType(
      nodeType,
      state,
      selectionsToAST([Array.from(nullthrows(nodeSelections).values())], state)
    );
  }
  if (schemaName === "__typename" && concreteType) {
    value = ts.createLiteralTypeNode(ts.createLiteral(concreteType));
  }
  return readOnlyObjectTypeProperty(key, value, conditional);
}

const isTypenameSelection = (selection: Selection) =>
  selection.schemaName === "__typename";
const hasTypenameSelection = (selections: Selection[]) =>
  selections.some(isTypenameSelection);
const onlySelectsTypename = (selections: Selection[]) =>
  selections.every(isTypenameSelection);

function selectionsToAST(
  selections: Selection[][],
  state: State,
  refTypeName?: string
): ts.TypeNode {
  const baseFields = new Map();
  const byConcreteType: { [type: string]: Selection[] } = {};

  flattenArray(selections).forEach(selection => {
    const { concreteType } = selection;
    if (concreteType) {
      byConcreteType[concreteType] = byConcreteType[concreteType] || [];
      byConcreteType[concreteType].push(selection);
    } else {
      const previousSel = baseFields.get(selection.key);

      baseFields.set(
        selection.key,
        previousSel ? mergeSelection(selection, previousSel) : selection
      );
    }
  });

  const types: ts.PropertySignature[][] = [];

  if (
    Object.keys(byConcreteType).length &&
    onlySelectsTypename(Array.from(baseFields.values())) &&
    (hasTypenameSelection(Array.from(baseFields.values())) ||
      Object.keys(byConcreteType).every(type =>
        hasTypenameSelection(byConcreteType[type])
      ))
  ) {
    for (const concreteType in byConcreteType) {
      types.push(
        groupRefs([
          ...Array.from(baseFields.values()),
          ...byConcreteType[concreteType]
        ]).map(selection => makeProp(selection, state, concreteType))
      );
    }
    // It might be some other type than the listed concrete types. Ideally, we
    // would set the type to diff(string, set of listed concrete types), but
    // this doesn't exist in Flow at the time.
    const otherProp = readOnlyObjectTypeProperty(
      "__typename",
      ts.createLiteralTypeNode(ts.createLiteral("%other"))
    );
    const otherPropWithComment = ts.addSyntheticLeadingComment(
      otherProp,
      ts.SyntaxKind.MultiLineCommentTrivia,
      "This will never be '% other', but we need some\n" +
        "value in case none of the concrete values match.",
      true
    );
    types.push([otherPropWithComment]);
  } else {
    let selectionMap = selectionsToMap(Array.from(baseFields.values()));
    for (const concreteType in byConcreteType) {
      selectionMap = mergeSelections(
        selectionMap,
        selectionsToMap(
          byConcreteType[concreteType].map(sel => ({
            ...sel,
            conditional: true
          }))
        )
      );
    }
    const selectionMapValues = groupRefs(Array.from(selectionMap.values())).map(
      sel =>
        isTypenameSelection(sel) && sel.concreteType
          ? makeProp({ ...sel, conditional: false }, state, sel.concreteType)
          : makeProp(sel, state)
    );
    types.push(selectionMapValues);
  }

  return ts.createUnionTypeNode(
    types.map(props => {
      if (refTypeName) {
        props.push(
          readOnlyObjectTypeProperty(
            " $refType",
            ts.createTypeReferenceNode(
              ts.createIdentifier(refTypeName),
              undefined
            )
          )
        );
      }
      return exactObjectTypeAnnotation(props);
    })
  );
}

// We don't have exact object types in typescript.
function exactObjectTypeAnnotation(
  properties: ts.PropertySignature[]
): ts.TypeLiteralNode {
  return ts.createTypeLiteralNode(properties);
}

const idRegex = /^[$a-zA-Z_][$a-z0-9A-Z_]*$/;

function readOnlyObjectTypeProperty(
  propertyName: string,
  type: ts.TypeNode,
  optional?: boolean
): ts.PropertySignature {
  return ts.createPropertySignature(
    [ts.createToken(ts.SyntaxKind.ReadonlyKeyword)],
    idRegex.test(propertyName)
      ? ts.createIdentifier(propertyName)
      : ts.createLiteral(propertyName),
    optional ? ts.createToken(ts.SyntaxKind.QuestionToken) : undefined,
    type,
    undefined
  );
}

function mergeSelection(
  a: Selection | null | undefined,
  b: Selection
): Selection {
  if (!a) {
    return {
      ...b,
      conditional: true
    };
  }
  return {
    ...a,
    nodeSelections: a.nodeSelections
      ? mergeSelections(a.nodeSelections, nullthrows(b.nodeSelections))
      : null,
    conditional: a.conditional && b.conditional
  };
}

function mergeSelections(a: SelectionMap, b: SelectionMap): SelectionMap {
  const merged = new Map();
  for (const [key, value] of Array.from(a.entries())) {
    merged.set(key, value);
  }
  for (const [key, value] of Array.from(b.entries())) {
    merged.set(key, mergeSelection(a.get(key), value));
  }
  return merged;
}

function isPlural(node: Fragment): boolean {
  return Boolean(node.metadata && node.metadata.plural);
}

function exportType(name: string, type: ts.TypeNode): ts.Statement {
  return ts.createTypeAliasDeclaration(
    undefined,
    [ts.createToken(ts.SyntaxKind.ExportKeyword)],
    ts.createIdentifier(name),
    undefined,
    type
  );
}

function importTypes(names: string[], fromModule: string): ts.Statement {
  return ts.createImportDeclaration(
    undefined,
    undefined,
    ts.createImportClause(
      undefined,
      ts.createNamedImports(
        names.map(name =>
          ts.createImportSpecifier(undefined, ts.createIdentifier(name))
        )
      )
    ),
    ts.createLiteral(fromModule)
  );
}

function createVisitor(options: TypeGeneratorOptions) {
  const state: State = {
    customScalars: options.customScalars,
    enumsHasteModule: options.enumsHasteModule,
    existingFragmentNames: options.existingFragmentNames,
    inputFieldWhiteList: options.inputFieldWhiteList,
    relayRuntimeModule: options.relayRuntimeModule,
    usedEnums: {},
    usedFragments: new Set(),
    usedObjectTypes: new Map(),
    useHaste: options.useHaste,
    useSingleArtifactDirectory: options.useSingleArtifactDirectory
  };

  return {
    leave: {
      Root(node: any) {
        const inputVariablesType = generateInputVariablesType(node, state);
        const responseType = exportType(
          `${node.name}Response`,
          selectionsToAST(node.selections, state)
        );
        return [
          // TODO: This is disabled until TS 2.8 is released which has the features we need to properly support fragment
          //       reference checking. See https://github.com/alloy/DefinitelyTyped/pull/1
          //
          // ...getFragmentImports(state),
          ...getEnumDefinitions(state),
          ...getObjectTypeDefinitions(state),
          inputVariablesType,
          responseType
        ];
      },

      Fragment(node: any) {
        const flattenedSelections: Selection[] = flattenArray(node.selections);
        const numConecreteSelections = flattenedSelections.filter(
          s => s.concreteType
        ).length;
        const selections = flattenedSelections.map(selection => {
          if (
            numConecreteSelections <= 1 &&
            isTypenameSelection(selection) &&
            !isAbstractType(node.type)
          ) {
            return [
              {
                ...selection,
                concreteType: node.type.toString()
              }
            ];
          }
          return [selection];
        });
        // TODO: This is disabled until TS 2.8 is released which has the features we need to properly support fragment
        //       reference checking. See https://github.com/alloy/DefinitelyTyped/pull/1
        //
        // const refTypeName = getRefTypeName(node.name);
        // const _refType = ts.createEnumDeclaration(
        //   undefined,
        //   [ts.createToken(ts.SyntaxKind.ConstKeyword)],
        //   ts.createIdentifier(`_${refTypeName}`),
        //   []
        // );
        // const refType = ts.createTypeAliasDeclaration(
        //   undefined,
        //   [ts.createToken(ts.SyntaxKind.ExportKeyword)],
        //   refTypeName,
        //   undefined,
        //   ts.createIntersectionTypeNode([
        //     ts.createTypeReferenceNode(_refType.name, undefined),
        //     ts.createTypeReferenceNode("FragmentReference", undefined)
        //   ])
        // );
        // const baseType = selectionsToAST(selections, state, refTypeName);
        const baseType = selectionsToAST(selections, state);
        const type = isPlural(node)
          ? ts.createTypeReferenceNode(ts.createIdentifier("ReadonlyArray"), [
              baseType
            ])
          : baseType;
        return [
          // ...getFragmentImports(state),
          ...getEnumDefinitions(state),
          // importTypes(["FragmentReference"], state.relayRuntimeModule),
          // _refType,
          // refType,
          exportType(node.name, type)
        ];
      },

      InlineFragment(node: any) {
        const typeCondition = node.typeCondition;
        return flattenArray(node.selections).map(typeSelection => {
          return isAbstractType(typeCondition)
            ? {
                ...typeSelection,
                conditional: true
              }
            : {
                ...typeSelection,
                concreteType: typeCondition.toString()
              };
        });
      },
      Condition(node: any) {
        return flattenArray(node.selections).map(selection => {
          return {
            ...selection,
            conditional: true
          };
        });
      },
      ScalarField(node: any) {
        return [
          {
            key: node.alias || node.name,
            schemaName: node.name,
            value: transformScalarType(node.type, state)
          }
        ];
      },
      LinkedField(node: any) {
        return [
          {
            key: node.alias || node.name,
            schemaName: node.name,
            nodeType: node.type,
            nodeSelections: selectionsToMap(flattenArray(node.selections))
          }
        ];
      },
      FragmentSpread(node: any) {
        state.usedFragments.add(node.name);
        return [
          {
            key: "__fragments_" + node.name,
            ref: node.name
          }
        ];
      }
    }
  };
}

function selectionsToMap(selections: Array<Selection>): SelectionMap {
  const map = new Map();
  selections.forEach(selection => {
    const previousSel = map.get(selection.key);
    map.set(
      selection.key,
      previousSel ? mergeSelection(previousSel, selection) : selection
    );
  });
  return map;
}

function flattenArray<T>(arrayOfArrays: T[][]): T[] {
  const result: T[] = [];
  arrayOfArrays.forEach(array => result.push(...array));
  return result;
}

function generateInputVariablesType(node: Root, state: State) {
  return exportType(
    `${node.name}Variables`,
    exactObjectTypeAnnotation(
      node.argumentDefinitions.map(arg => {
        return readOnlyObjectTypeProperty(
          arg.name,
          transformInputType(arg.type, state),
          !(arg.type instanceof GraphQLNonNull)
        );
      })
    )
  );
}

function groupRefs(props: Selection[]): Selection[] {
  const result: Selection[] = [];
  const refs: string[] = [];
  props.forEach(prop => {
    if (prop.ref) {
      refs.push(prop.ref);
    } else {
      result.push(prop);
    }
  });
  if (refs.length > 0) {
    const value = ts.createIntersectionTypeNode(
      refs.map(ref =>
        ts.createTypeReferenceNode(
          ts.createIdentifier(getRefTypeName(ref)),
          undefined
        )
      )
    );
    // TODO: This is disabled until TS 2.8 is released which has the features we need to properly support fragment
    //       reference checking. See https://github.com/alloy/DefinitelyTyped/pull/1
    //
    // result.push({
    //   key: " $fragmentRefs",
    //   conditional: false,
    //   value
    // });
  }
  return result;
}

function createAnyTypeAlias(name: string): ts.TypeAliasDeclaration {
  return ts.createTypeAliasDeclaration(
    undefined,
    undefined,
    ts.createIdentifier(name),
    undefined,
    ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
  );
}

function getFragmentImports(state: State) {
  const imports: ts.Statement[] = [];
  if (state.usedFragments.size > 0) {
    const usedFragments = Array.from(state.usedFragments).sort();
    for (const usedFragment of usedFragments) {
      const refTypeName = getRefTypeName(usedFragment);
      if (
        state.useSingleArtifactDirectory &&
        state.existingFragmentNames.has(usedFragment)
      ) {
        imports.push(importTypes([refTypeName], `./${usedFragment}.graphql`));
      } else {
        imports.push(createAnyTypeAlias(refTypeName));
      }
    }
  }
  return imports;
}

function anyTypeAlias(typeName: string): ts.Statement {
  return ts.createTypeAliasDeclaration(
    undefined,
    undefined,
    ts.createIdentifier(typeName),
    undefined,
    ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
  );
}

function getObjectTypeDefinitions(state: State): ts.Statement[] {
  const { usedObjectTypes } = state;
  const typeDefs: ts.Statement[] = [];

  usedObjectTypes.forEach(t => t && typeDefs.push(t));
  return typeDefs;
}

function getEnumDefinitions({ enumsHasteModule, usedEnums }: State) {
  const enumNames = Object.keys(usedEnums).sort();
  if (enumNames.length === 0) {
    return [];
  }
  if (enumsHasteModule) {
    return [importTypes(enumNames, enumsHasteModule)];
  }
  return enumNames.map(name => {
    const values = usedEnums[name].getValues().map(({ value }) => value);
    values.sort();
    values.push("%future added value");
    return exportType(
      name,
      ts.createUnionTypeNode(
        values.map(value => stringLiteralTypeAnnotation(value))
      )
    );
  });
}

function stringLiteralTypeAnnotation(name: string): ts.TypeNode {
  return ts.createLiteralTypeNode(ts.createLiteral(name));
}

function getRefTypeName(name: string): string {
  return `${name}$ref`;
}

export const transforms: TypeGenerator["transforms"] = [
  IRTransforms.commonTransforms[2], // RelayRelayDirectiveTransform.transform
  IRTransforms.commonTransforms[3], // RelayMaskTransform.transform
  IRTransforms.printTransforms[0] // FlattenTransform.transformWithOptions({})
];
