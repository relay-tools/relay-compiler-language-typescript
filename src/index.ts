import { PluginInterface } from 'relay-compiler';

import { formatGeneratedModule } from './formatGeneratedModule';
import { find } from './FindGraphQLTags';
import * as TypeScriptGenerator from './TypeScriptGenerator';

export default function plugin(): PluginInterface {
  return {
    inputExtensions: ['ts', 'tsx'],
    outputExtension: 'ts',
    findGraphQLTags: find,
    formatModule: formatGeneratedModule,
    typeGenerator: TypeScriptGenerator,
  }
}
