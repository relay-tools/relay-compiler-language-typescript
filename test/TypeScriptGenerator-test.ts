// @ts-ignore
// TODO - Update CompilerContext type when release @types/relay-compiler 8
import { CompilerContext, IRTransforms, Root } from "relay-compiler";
import { TypeGeneratorOptions } from "relay-compiler/lib/language/RelayLanguagePluginInterface";
import { generateTestsFromFixtures } from "relay-test-utils-internal/lib/generateTestsFromFixtures";
import * as parseGraphQLText from "relay-test-utils-internal/lib/parseGraphQLText";
import { TestSchema } from "relay-test-utils-internal/lib/TestSchema";
import * as TypeScriptGenerator from "../src/TypeScriptGenerator";

function generate(
  text,
  options: TypeGeneratorOptions,
  // TODO - Update CompilerContext type when release @types/relay-compiler 8
  context?
) {
  const schema = TestSchema.extend([
    ...IRTransforms.schemaExtensions,
    `
      scalar Color
      extend type User {
        color: Color
      }
      type ExtraUser implements Actor {
        address: StreetAddress
        allPhones: [Phone]
        birthdate: Date
        emailAddresses: [String]
        firstName(if: Boolean, unless: Boolean): String
        friends(after: ID, before: ID, first: Int, last: Int, orderby: [String], find: String, isViewerFriend: Boolean, if: Boolean, unless: Boolean, traits: [PersonalityTraits]): FriendsConnection
        hometown: Page
        id: ID!
        lastName: String
        name: String
        nameRenderer(supported: [String!]): UserNameRenderer
        nameRenderable(supported: [String!]): UserNameRenderable
        profilePicture(size: [Int], preset: PhotoSize): Image
        screennames: [Screenname]
        subscribeStatus: String
        subscribers(first: Int): SubscribersConnection
        url(relative: Boolean, site: String): String
        websites: [String]
        username: String
      }
    `
  ]);
  const { definitions, schema: extendedSchema } = parseGraphQLText(
    schema,
    text
  );
  return new CompilerContext(extendedSchema)
    .addAll(definitions)
    .applyTransforms(TypeScriptGenerator.transforms)
    .documents()
    .map(
      doc =>
        `// ${doc.name}.graphql\n${TypeScriptGenerator.generate(
          extendedSchema,
          doc as any,
          {
            ...options,
            normalizationIR: context
              ? (context.get(doc.name) as Root)
              : undefined
          }
        )}`
    )
    .join("\n\n");
}

describe("Snapshot tests", () => {
  function generateContext(text) {
    const relaySchema = TestSchema.extend(IRTransforms.schemaExtensions);
    const { definitions, schema: extendedSchema } = parseGraphQLText(
      relaySchema,
      text
    );
    return new CompilerContext(extendedSchema)
      .addAll(definitions)
      .applyTransforms([
        ...IRTransforms.commonTransforms,
        ...IRTransforms.queryTransforms,
        ...IRTransforms.codegenTransforms
      ]);
  }

  describe("TypeScriptGenerator with a single artifact directory", () => {
    generateTestsFromFixtures(`${__dirname}/fixtures/type-generator`, text => {
      const context = generateContext(text);
      return generate(
        text,
        {
          customScalars: {},
          // enumsHasteModule: null,
          existingFragmentNames: new Set(["PhotoFragment"]),
          optionalInputFields: [],
          useHaste: false,
          useSingleArtifactDirectory: true,
          noFutureProofEnums: false
        },
        context
      );
    });
  });

  describe("TypeScriptGenerator without a single artifact directory", () => {
    generateTestsFromFixtures(`${__dirname}/fixtures/type-generator`, text => {
      const context = generateContext(text);
      return generate(
        text,
        {
          customScalars: {},
          // enumsHasteModule: null,
          existingFragmentNames: new Set(["PhotoFragment"]),
          optionalInputFields: [],
          useHaste: false,
          useSingleArtifactDirectory: false,
          noFutureProofEnums: false
        },
        context
      );
    });
  });
});

describe("Does not add `%future added values` when the noFutureProofEnums option is set", () => {
  const text = `
    fragment ScalarField on User {
      traits
    }
  `;
  const types = generate(text, {
    customScalars: {},
    // enumsHasteModule: null,
    existingFragmentNames: new Set(["PhotoFragment"]),
    optionalInputFields: [],
    useHaste: true,
    useSingleArtifactDirectory: false,
    // This is what's different from the tests above.
    noFutureProofEnums: true
  });

  // Without the option, PersonalityTraits would be `"CHEERFUL" | ... | "%future added value");`
  expect(types).toContain(
    'export type PersonalityTraits = "CHEERFUL" | "DERISIVE" | "HELPFUL" | "SNARKY";'
  );
});

describe.each`
  mapping     | type
  ${"String"} | ${"string"}
  ${"Url"}    | ${"string"}
  ${"ID"}     | ${"string"}
  ${"Int"}    | ${"number"}
  ${"Color"}  | ${"Color"}
  ${"{}"}     | ${"{}"}
  ${"[]"}     | ${"[]"}
`("Custom scalar mapping $mapping to $type", ({ mapping, type }) => {
  const text = `
    fragment Test on User {
        color
    }
  `;
  const types = generate(text, {
    customScalars: {
      Color: mapping
    },
    // enumsHasteModule: null,
    existingFragmentNames: new Set(["PhotoFragment"]),
    optionalInputFields: [],
    useHaste: false,
    useSingleArtifactDirectory: true,
    noFutureProofEnums: false
  });

  expect(types).toContain(`color: ${type} | null`);
});
