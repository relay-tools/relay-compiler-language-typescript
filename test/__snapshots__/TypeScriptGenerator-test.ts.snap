// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScriptGenerator matches expected output: conditional.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment ConditionField on Node {
  id @include(if: $condition)
}

fragment NestedCondition on Node {
  ... @include(if: $condition) {
    id
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type ConditionField = {
    readonly id?: string;
};


export type NestedCondition = {
    readonly id?: string;
};

`;

exports[`TypeScriptGenerator matches expected output: fragment-spread.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment FragmentSpread on Node {
  id
  ...OtherFragment
  justFrag: profilePicture {
    ...PictureFragment
  }
  fragAndField: profilePicture {
    uri
    ...PictureFragment
  }
  ... on User {
    ...UserFrag1
    ...UserFrag2
  }
}

fragment ConcreateTypes on Viewer {
  actor {
    __typename
    ... on Page {
      id
      ...PageFragment
    }
    ... on User {
      name
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type FragmentSpread = {
    readonly id: string;
    readonly justFrag: ({
    }) | null;
    readonly fragAndField: ({
        readonly uri: string | null;
    }) | null;
};


export type ConcreateTypes = {
    readonly actor: ({
        readonly __typename: "Page";
        readonly id: string;
    } | {
        readonly __typename: "User";
        readonly name: string | null;
    } | {
        /*This will never be '% other', but we need some
        value in case none of the concrete values match.*/
        readonly __typename: "%other";
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: inline-fragment.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment InlineFragment on Node {
  id
  ... on Actor {
    id
    name
  }
  ... on User {
    message {
      text
    }
  }
}

fragment InlineFragmentWithOverlappingFields on Actor {
  ... on User {
    hometown {
      id
      name
    }
  }
  ... on Page {
    name
    hometown {
      id
      message {
        text
      }
    }
  }
}

fragment InlineFragmentConditionalID on Node {
  ... on Actor {
    id # nullable since it's conditional
    name
  }
}

fragment InlineFragmentKitchenSink on Story {
  actor {
    id
    profilePicture {
      uri
    }
    ... on User {
      id
      name
      ...SomeFragment
      profilePicture {
        width
      }
    }
    ... on Page {
      profilePicture {
        uri
        height
      }
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type InlineFragment = {
    readonly id: string;
    readonly name?: string | null;
    readonly message?: ({
        readonly text: string | null;
    }) | null;
};


export type InlineFragmentWithOverlappingFields = {
    readonly hometown?: ({
        readonly id: string;
        readonly name: string | null;
        readonly message?: ({
            readonly text: string | null;
        }) | null;
    }) | null;
    readonly name?: string | null;
};


export type InlineFragmentConditionalID = {
    readonly id?: string;
    readonly name?: string | null;
};


export type InlineFragmentKitchenSink = {
    readonly actor: ({
        readonly id: string;
        readonly profilePicture: ({
            readonly uri: string | null;
            readonly width?: number | null;
            readonly height?: number | null;
        }) | null;
        readonly name?: string | null;
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: linked-field.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment LinkedField on User {
  profilePicture {
    uri
    width
    height
  }
  hometown {
    # object
    id
    profilePicture {
      uri
    }
  }
  actor {
    # interface
    id
  }
}

query UnionTypeTest {
  neverNode {
    __typename
    ... on FakeNode {
      id
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type LinkedField = {
    readonly profilePicture: ({
        readonly uri: string | null;
        readonly width: number | null;
        readonly height: number | null;
    }) | null;
    readonly hometown: ({
        readonly id: string;
        readonly profilePicture: ({
            readonly uri: string | null;
        }) | null;
    }) | null;
    readonly actor: ({
        readonly id: string;
    }) | null;
};


export type UnionTypeTestVariables = {
};
export type UnionTypeTestResponse = {
    readonly neverNode: ({
        readonly __typename: "FakeNode";
        readonly id: string;
    } | {
        /*This will never be '% other', but we need some
        value in case none of the concrete values match.*/
        readonly __typename: "%other";
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: mutation.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
mutation CommentCreateMutation(
  $input: CommentCreateInput!
  $first: Int
  $orderBy: [String!]
) {
  commentCreate(input: $input) {
    comment {
      id
      name
      friends(first: $first, orderby: $orderBy) {
        count
      }
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export interface CommentCreateInput {
    readonly clientMutationId: string | null;
    readonly feedbackId: string | null;
}
export type CommentCreateMutationVariables = {
    readonly input: CommentCreateInput;
    readonly first?: number | null;
    readonly orderBy?: ReadonlyArray<string> | null;
};
export type CommentCreateMutationResponse = {
    readonly commentCreate: ({
        readonly comment: ({
            readonly id: string;
            readonly name: string | null;
            readonly friends: ({
                readonly count: number | null;
            }) | null;
        }) | null;
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: mutation-input-has-array.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
mutation InputHasArray($input: UpdateAllSeenStateInput) {
  viewerNotificationsUpdateAllSeenState(input: $input) {
    stories {
      actorCount
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export interface UpdateAllSeenStateInput {
    readonly clientMutationId: string | null;
    readonly storyIds: ReadonlyArray<string | null> | null;
}
export type InputHasArrayVariables = {
    readonly input?: UpdateAllSeenStateInput | null;
};
export type InputHasArrayResponse = {
    readonly viewerNotificationsUpdateAllSeenState: ({
        readonly stories: ReadonlyArray<({
                readonly actorCount: number | null;
            }) | null> | null;
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: plural-fragment.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment PluralFragment on Node @relay(plural: true) {
  id
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type PluralFragment = ReadonlyArray<{
        readonly id: string;
    }>;

`;

exports[`TypeScriptGenerator matches expected output: roots.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
query ExampleQuery($id: ID!) {
  node(id: $id) {
    id
  }
}

fragment ExampleFragment on User {
  id
}

mutation TestMutation($input: CommentCreateInput!) {
  commentCreate(input: $input) {
    comment {
      id
    }
  }
}

subscription TestSubscription($input: FeedbackLikeInput) {
  feedbackLikeSubscribe(input: $input) {
    feedback {
      id
    }
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type ExampleQueryVariables = {
    readonly id: string;
};
export type ExampleQueryResponse = {
    readonly node: ({
        readonly id: string;
    }) | null;
};


export type ExampleFragment = {
    readonly id: string;
};


export interface CommentCreateInput {
    readonly clientMutationId: string | null;
    readonly feedbackId: string | null;
}
export type TestMutationVariables = {
    readonly input: CommentCreateInput;
};
export type TestMutationResponse = {
    readonly commentCreate: ({
        readonly comment: ({
            readonly id: string;
        }) | null;
    }) | null;
};


export interface FeedbackLikeInput {
    readonly clientMutationId: string | null;
    readonly feedbackId: string | null;
}
export type TestSubscriptionVariables = {
    readonly input?: FeedbackLikeInput | null;
};
export type TestSubscriptionResponse = {
    readonly feedbackLikeSubscribe: ({
        readonly feedback: ({
            readonly id: string;
        }) | null;
    }) | null;
};

`;

exports[`TypeScriptGenerator matches expected output: scalar-field.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment ScalarField on User {
  id
  name
  websites
  traits
  aliasedLinkedField: birthdate {
    aliasedField: year
  }
  screennames {
    name
    service
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type PersonalityTraits = "CHEERFUL" | "DERISIVE" | "HELPFUL" | "SNARKY" | "%future added value";
export type ScalarField = {
    readonly id: string;
    readonly name: string | null;
    readonly websites: ReadonlyArray<string | null> | null;
    readonly traits: ReadonlyArray<PersonalityTraits | null> | null;
    readonly aliasedLinkedField: ({
        readonly aliasedField: number | null;
    }) | null;
    readonly screennames: ReadonlyArray<({
            readonly name: string | null;
            readonly service: string | null;
        }) | null> | null;
};

`;

exports[`TypeScriptGenerator matches expected output: typename-on-union.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment TypenameInside on Actor {
  ... on User {
    __typename
    firstName
  }
  ... on Page {
    __typename
    username
  }
}

fragment TypenameOutside on Actor {
  __typename
  ... on User {
    firstName
  }
  ... on Page {
    username
  }
}

fragment TypenameOutsideWithAbstractType on Node {
  __typename
  ... on User {
    firstName
    address {
      street # only here
      city # common
    }
  }
  ... on Actor {
    username
    address {
      city # common
      country # only here
    }
  }
}

fragment TypenameWithoutSpreads on User {
  __typename
  firstName
}

fragment TypenameWithoutSpreadsAbstractType on Node {
  __typename
  id
}


fragment TypenameWithCommonSelections on Actor {
  __typename
  name
  ... on User {
    firstName
  }
  ... on Page {
    username
  }
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type TypenameInside = {
    readonly __typename: "User";
    readonly firstName: string | null;
} | {
    readonly __typename: "Page";
    readonly username: string | null;
} | {
    /*This will never be '% other', but we need some
    value in case none of the concrete values match.*/
    readonly __typename: "%other";
};


export type TypenameOutside = {
    readonly __typename: "User";
    readonly firstName: string | null;
} | {
    readonly __typename: "Page";
    readonly username: string | null;
} | {
    /*This will never be '% other', but we need some
    value in case none of the concrete values match.*/
    readonly __typename: "%other";
};


export type TypenameOutsideWithAbstractType = {
    readonly __typename: string;
    readonly username?: string | null;
    readonly address?: ({
        readonly city: string | null;
        readonly country: string | null;
        readonly street?: string | null;
    }) | null;
    readonly firstName?: string | null;
};


export type TypenameWithoutSpreads = {
    readonly firstName: string | null;
    readonly __typename: "User";
};


export type TypenameWithoutSpreadsAbstractType = {
    readonly __typename: string;
    readonly id: string;
};


export type TypenameWithCommonSelections = {
    readonly __typename: string;
    readonly name: string | null;
    readonly firstName?: string | null;
    readonly username?: string | null;
};

`;

exports[`TypeScriptGenerator matches expected output: unmasked-fragment-spreads.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment UserProfile on User {
  profilePicture(size: $ProfilePicture_SIZE) {
    ...PhotoFragment @relay(mask: false)

    # duplicated field should be merged
    ...AnotherRecursiveFragment @relay(mask: false)

    # Compose child fragment
    ...PhotoFragment
  }
}

fragment PhotoFragment on Image {
  uri
  ...RecursiveFragment @relay(mask: false)
}

fragment RecursiveFragment on Image {
  uri
  width
}

fragment AnotherRecursiveFragment on Image {
  uri
  height
}

~~~~~~~~~~ OUTPUT ~~~~~~~~~~
export type UserProfile = {
    readonly profilePicture: ({
        readonly uri: string | null;
        readonly width: number | null;
        readonly height: number | null;
    }) | null;
};


export type PhotoFragment = {
    readonly uri: string | null;
    readonly width: number | null;
};


export type RecursiveFragment = {
    readonly uri: string | null;
    readonly width: number | null;
};


export type AnotherRecursiveFragment = {
    readonly uri: string | null;
    readonly height: number | null;
};

`;
